src/ploneintranet/workspace/browser/tiles/sidebar.py

```diff
@@ -127,22 +136,60 @@ def users(self):
     def existing_users(self):
         return self.workspace().existing_users()
 
-    def __call__(self):
+    def execute_batch_function(self):
         form = self.request.form
         ws = self.workspace()
-        user_id = form.get('user_id')
-        if user_id:
-            if not self.can_manage_workspace():
+        user_ids = form.get('user_id')
+        if isinstance(user_ids, basestring):
+            user_ids = user_ids.split(',')
+        batch_function = form.get('batch-function')
+        if user_ids:
+            if not self.can_manage_roster():
                 msg = _(u'You do not have permission to change the workspace '
                         u'policy')
                 raise Unauthorized(msg)
             else:
-                IWorkspace(ws).add_to_team(user=user_id)
-                api.portal.show_message(
-                    _(u'Member added'),
-                    self.request,
-                    'success',
-                )
+                if batch_function == 'add':
+                    for user_id in user_ids:
+                        IWorkspace(ws).add_to_team(user=user_id)
+                    api.portal.show_message(
+                        _(u'Member(s) added'),
+                        self.request,
+                        'success',
+                    )
+                elif batch_function == 'remove':
+                    for user_id in user_ids:
+                        IWorkspace(ws).remove_from_team(user=user_id)
+                    api.portal.show_message(
+                        _(u'Member(s) removed'),
+                        self.request,
+                        'success',
+                    )
+                elif batch_function == 'role':
+                    role = self.request.get('role')
+                    default_role = ws.participant_policy.title()
+                    groups = [default_role]
+                    if role:
+                        groups.append(role)
```

Comment:
This is in line with our original technical design, but it conflicts with the UI design. This adds custom roles on top of the default workspace roles. The UI clearly indicates that you're reassigning the user from the default role to another role.

In this implementation, if you're in a 'publish' workspace you can move users to 'consume' visually but it has no security effect whatsoever, because the user is in both ('publish', 'consume') groups and publish is a superset of consume.

The reason that this problem doesn't show up in the tests, is because they are run in a consume workspace and elevate permissions.

To summarize: with this implementation you can only elevate permissions, not assign lesser permissions than the default. IMO that should be changed so that a custom role *replaces* rather than *adds to* the default group.

@mattss do you agree?